diff --git a/src/vulkan/wrapper/wrapper_device.c b/src/vulkan/wrapper/wrapper_device.c
index 2ea37d1b392..dd4a877e42e 100644
--- a/src/vulkan/wrapper/wrapper_device.c
+++ b/src/vulkan/wrapper/wrapper_device.c
@@ -138,6 +138,7 @@ wrapper_CreateDevice(VkPhysicalDevice physicalDevice,
                      const VkAllocationCallbacks* pAllocator,
                      VkDevice* pDevice)
 {
+   __log("wrapper_CreateDevice");
    VK_FROM_HANDLE(wrapper_physical_device, physical_device, physicalDevice);
    const char *wrapper_enable_extensions[VK_DEVICE_EXTENSION_COUNT];
    uint32_t wrapper_enable_extension_count = 0;
diff --git a/src/vulkan/wrapper/wrapper_device_memory.c b/src/vulkan/wrapper/wrapper_device_memory.c
index 6d47a52fd18..1a9cb9bf068 100644
--- a/src/vulkan/wrapper/wrapper_device_memory.c
+++ b/src/vulkan/wrapper/wrapper_device_memory.c
@@ -1,10 +1,10 @@
-#define native_handle_t __native_handle_t
-#define buffer_handle_t __buffer_handle_t
+// #define native_handle_t __native_handle_t
+// #define buffer_handle_t __buffer_handle_t
 #include "wrapper_private.h"
 #include "wrapper_entrypoints.h"
 #include "vk_common_entrypoints.h"
-#undef native_handle_t
-#undef buffer_handle_t
+// #undef native_handle_t
+// #undef buffer_handle_t
 #include "util/os_file.h"
 #include "vk_util.h"
 
diff --git a/src/vulkan/wrapper/wrapper_instance.c b/src/vulkan/wrapper/wrapper_instance.c
index 237108ca6eb..eb85cf265ff 100644
--- a/src/vulkan/wrapper/wrapper_instance.c
+++ b/src/vulkan/wrapper/wrapper_instance.c
@@ -279,6 +279,8 @@ VKAPI_ATTR PFN_vkVoidFunction VKAPI_CALL
 wrapper_GetInstanceProcAddr(VkInstance _instance,
                             const char *pName)
 {
+
+   __log("vk_icdGetInstanceProcAddr %s", pName);
    VK_FROM_HANDLE(wrapper_instance, instance, _instance);
    return vk_instance_get_proc_addr(&instance->vk,
                                     &wrapper_instance_entrypoints,
diff --git a/src/vulkan/wrapper/wrapper_physical_device.c b/src/vulkan/wrapper/wrapper_physical_device.c
index 384c45483c6..14942851c4b 100644
--- a/src/vulkan/wrapper/wrapper_physical_device.c
+++ b/src/vulkan/wrapper/wrapper_physical_device.c
@@ -12,8 +12,16 @@
 #include "wsi_common.h"
 #include "util/os_misc.h"
 
+int __android_log_print(
+  int prio,
+  const char *tag,
+  const char *fmt,
+  ...
+);
+
 static VkResult
 wrapper_setup_device_extensions(struct wrapper_physical_device *pdevice) {
+   __android_log_print(6, "Wrapper", "inside wrapper_setup_device_extensions");
    struct vk_device_extension_table *exts = &pdevice->vk.supported_extensions;
    VkExtensionProperties pdevice_extensions[VK_DEVICE_EXTENSION_COUNT];
    uint32_t pdevice_extension_count = VK_DEVICE_EXTENSION_COUNT;
diff --git a/src/vulkan/wrapper/wrapper_private.h b/src/vulkan/wrapper/wrapper_private.h
index 7f8cc6e8c84..6b6ba9a9991 100644
--- a/src/vulkan/wrapper/wrapper_private.h
+++ b/src/vulkan/wrapper/wrapper_private.h
@@ -15,6 +15,23 @@ extern const struct vk_instance_extension_table wrapper_instance_extensions;
 extern const struct vk_device_extension_table wrapper_device_extensions;
 extern const struct vk_device_extension_table wrapper_filter_extensions;
 
+static void __log(const char* fmt, ...) {
+   static FILE* fd;
+   static pid_t pid;
+   if (fd == NULL) {
+      pid = getpid();
+      char buf[256];
+      sprintf(buf, "/sdcard/Documents/wrapper_%d.txt", pid);
+      fd = fopen(buf, "w");
+   }
+   // fprintf(fd, "[%d] ", pid);
+   va_list args;
+   va_start(args, fmt);
+   vfprintf(fd, fmt, args);
+   va_end(args);
+   fprintf(fd, "\n");
+}
+
 struct wrapper_instance {
    struct vk_instance vk;
 
diff --git a/src/vulkan/wsi/wsi_common_x11.c b/src/vulkan/wsi/wsi_common_x11.c
index 51a11b875ad..39b1faa18b7 100644
--- a/src/vulkan/wsi/wsi_common_x11.c
+++ b/src/vulkan/wsi/wsi_common_x11.c
@@ -64,13 +64,13 @@
 
 #include <android/hardware_buffer.h>
 
-typedef struct
-{
-    int version;        /* sizeof(native_handle) */
-    int numFds;         /* number of file-descriptors at &data[0] */
-    int numInts;        /* number of ints at &data[numFds] */
-    int data[0];        /* numFds + numInts ints */
-} native_handle_t;
+// typedef struct native_handle
+// {
+//     int version;        /* sizeof(native_handle) */
+//     int numFds;         /* number of file-descriptors at &data[0] */
+//     int numInts;        /* number of ints at &data[numFds] */
+//     int data[0];        /* numFds + numInts ints */
+// } native_handle_t;
 
 const native_handle_t* _Nullable AHardwareBuffer_getNativeHandle(
         const AHardwareBuffer* _Nonnull buffer);
