diff --git a/include/android_stub/cutils/log.h b/include/android_stub/cutils/log.h
index 0e0248e50fa..51a000e1f79 100644
--- a/include/android_stub/cutils/log.h
+++ b/include/android_stub/cutils/log.h
@@ -1 +1,3 @@
 #include <log/log.h>
+
+// Hello
\ No newline at end of file
diff --git a/src/vulkan/wrapper/meson.build b/src/vulkan/wrapper/meson.build
index 5b818b3b4bd..3337202b727 100644
--- a/src/vulkan/wrapper/meson.build
+++ b/src/vulkan/wrapper/meson.build
@@ -79,6 +79,8 @@ wrapper_files = files(
   'wrapper_device_memory.c',
   'wrapper_instance.c',
   'wrapper_physical_device.c',
+  'wrapper_image.c',
+  'wrapper_image_view.c',
 )
 
 wrapper_deps = [
diff --git a/src/vulkan/wrapper/vk_entrypoints.py b/src/vulkan/wrapper/vk_entrypoints.py
deleted file mode 120000
index bbc9deb820f..00000000000
--- a/src/vulkan/wrapper/vk_entrypoints.py
+++ /dev/null
@@ -1 +0,0 @@
-../util/vk_entrypoints.py
\ No newline at end of file
diff --git a/src/vulkan/wrapper/vk_entrypoints.py b/src/vulkan/wrapper/vk_entrypoints.py
new file mode 100644
index 00000000000..82974e06ec3
--- /dev/null
+++ b/src/vulkan/wrapper/vk_entrypoints.py
@@ -0,0 +1,431 @@
+# Copyright 2020 Intel Corporation
+#
+# Permission is hereby granted, free of charge, to any person obtaining a
+# copy of this software and associated documentation files (the
+# "Software"), to deal in the Software without restriction, including
+# without limitation the rights to use, copy, modify, merge, publish,
+# distribute, sub license, and/or sell copies of the Software, and to
+# permit persons to whom the Software is furnished to do so, subject to
+# the following conditions:
+#
+# The above copyright notice and this permission notice (including the
+# next paragraph) shall be included in all copies or substantial portions
+# of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT.
+# IN NO EVENT SHALL VMWARE AND/OR ITS SUPPLIERS BE LIABLE FOR
+# ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+import xml.etree.ElementTree as et
+import string
+
+from collections import OrderedDict, namedtuple, defaultdict
+
+# Mesa-local imports must be declared in meson variable
+# '{file_without_suffix}_depend_files'.
+from vk_extensions import get_all_required, filter_api
+
+EntrypointParam = namedtuple('EntrypointParam', 'type name decl len')
+
+WrappedStructs = {
+    "VkPhysicalDevice": "wrapper_physical_device",
+    "VkDevice": "wrapper_device",
+    "VkCommandBuffer": "wrapper_command_buffer",
+    "VkQueue": "wrapper_queue",
+    "VkImage": "wrapper_image",
+    # PASS
+}
+
+WrappedStructCounts = {
+    "VkPhysicalDevice": "physicalDeviceCount",
+    "VkDevice": "deviceCount",
+    "VkCommandBuffer": "commandBufferCount",
+    "VkQueue": "queueCount",
+    "VkImage": "imageCount",
+    # PASS
+}
+
+SKIP = ["VkExportMetalObjectsInfoEXT", "VkExportMetalIOSurfaceInfoEXT", "VkExportMetalTextureInfoEXT", "VkExportMetalCommandQueueInfoEXT"]
+TYPES = {}
+TYPE_DEPENDENCIES = {}
+PNEXT_DEPENDENCIES = {}
+PNEXT_MEMBERS = defaultdict(list)
+
+
+def get_types():
+    return set(TYPE_DEPENDENCIES) | set(PNEXT_DEPENDENCIES)
+
+class VkMember:
+    def __init__(self, type, name, optional):
+        self.type = VkType(type)
+        self.name = name
+        self.optional = optional
+        self.suffix = "_Array" if type in WrappedStructs and name.startswith("p") and name.endswith("s") else ""
+        self.length = f", payload->{self.count()}" if self.suffix else ""
+
+    def __repr__(self):
+        return f"{self.type} {self.name}" + ("[o]" if self.optional else "")
+
+    def count(self):
+        type = self.type.type
+        return WrappedStructCounts[type]
+
+class VkType:
+    def __init__(self, type):
+        self.type = type
+
+    def has_dependencies(self):
+        return self.type in TYPE_DEPENDENCIES or self.type in PNEXT_DEPENDENCIES
+
+    def members(self):
+        if self.type not in TYPES:
+            return []
+        members = []
+        for member in TYPES[self.type].findall("./member"):
+            t = member.find('./type').text
+            if t in TYPE_DEPENDENCIES[self.type]:
+                name = member.find('./name').text
+                optional = 'optional' in member.attrib and member.attrib['optional'] == 'true'
+                members.append(VkMember(t, name, optional))
+        return members
+
+    def pnexts(self):
+        return [VkType(x) for x in PNEXT_DEPENDENCIES[self.type]]
+
+    def sType(self):
+        if self.type not in TYPES:
+            return []
+        type = TYPES[self.type]
+        member = type.find("./member")
+        if not member:
+            return
+        if member.find('./name').text == 'sType' and 'values' in member.attrib:
+            return member.attrib['values']
+
+    def __repr__(self):
+        return self.type
+
+
+def generate_unwrapper_leaves():
+    # WrappedStructs
+    first = "\n".join([f"static {t} unwrap_{t}({t} payload) {{ VK_FROM_HANDLE({WrappedStructs[t]}, base, payload); return base->dispatch_handle; }}" for t in WrappedStructs])
+    second = "\n".join([
+        f"""
+        static const {t}* unwrap_{t}_Array(const {t}* payloads, uint32_t count) {{ 
+            // VK_FROM_HANDLE({WrappedStructs[t]}, base, payload); 
+            return payloads; 
+        }}
+        """
+        for t in WrappedStructs])
+    return first + second
+
+def generate_unwrapper_proto(vk_type):
+    return f"static {vk_type}* unwrap_{vk_type}(const {vk_type}* payload);"
+
+def generate_unwrapper(vk_type):
+    vk_type = VkType(vk_type)
+
+    template = string.Template("""
+static $X* unwrap_$X(const $X* payload) {
+    $X* copy = malloc(sizeof($X));
+    memcpy(copy, payload, sizeof($X));
+    $unwrap_members
+$while_loop
+    return copy;
+}
+    """)
+
+    while_loop_ = string.Template("""
+    // Deal with pnext
+    VkBaseOutStructure** prev = (VkBaseOutStructure**) &copy->pNext;
+    while (*prev) {
+        VkBaseOutStructure* current = (VkBaseOutStructure*) *prev;
+        $unwrap_pnext
+        prev = &current->pNext;
+    }""")
+
+    switch_ = string.Template("""
+        if (current->sType == $sType) {
+            current = (VkBaseOutStructure*) unwrap_$Z(($Z*) current);
+            *prev = current;
+        }
+    """)
+    if vk_type.has_dependencies():
+        print(vk_type.type)
+        for member in vk_type.members():
+            print(f"  unwrap_{member.type}({member.name})")
+        print("  members:", vk_type.members())
+        print("  pNext:", vk_type.pnexts())
+        while_loop = ""
+        if vk_type.pnexts():
+            while_loop = while_loop_.substitute(
+                unwrap_pnext="\n         ".join([switch_.substitute(Z=z, sType=z.sType()) for z in vk_type.pnexts()]))
+        return template.substitute(
+            X=vk_type,
+            unwrap_members="\n    ".join([
+                f"copy->{m.name} = unwrap_{m.type}{m.suffix}(payload->{m.name}{m.length});" 
+                for m in vk_type.members() if m.name != "physicalDevices"]),
+            while_loop=while_loop,
+        )
+    else:
+        return ""
+
+def generate_free(vk_type):
+    pass
+
+
+# Main template for a single trampoline function
+TRAMPOLINE_TEMPLATE = string.Template("""
+static VKAPI_ATTR $return_type VKAPI_CALL
+wrapper_tramp_$name(
+    $decl_params)
+{
+$handle_unwrap_logic
+$return_block $dispatch_table.$name($call_params);
+}""")
+
+
+def _generate_trampoline(command, dispatch_table="device->dispatch_table"):
+    """
+    Generates a complete C trampoline function for a given Vulkan command.
+    """
+    params = list(command.params)
+    handle_unwrap_logic = []
+    call = []
+
+    handle_unwrap_logic.append(f"    VK_FROM_HANDLE({WrappedStructs[params[0].type]}, base, {params[0].name});")
+    call.append(f"base->dispatch_handle")
+    for param in params[1:]:
+        # Only for non pointer types (which are output pointers), we don't get this info directly,
+        # but we can detect this with the name
+        if param.type in WrappedStructs and not param.name.startswith("p"):
+            handle_unwrap_logic.append(f"    VK_FROM_HANDLE({WrappedStructs[param.type]}, w_{param.name}, {param.name});")
+            call.append(f"w_{param.name}->dispatch_handle")
+        elif param.type in get_types():
+            handle_unwrap_logic.append(f"    {param.type}* w_{param.name} = unwrap_{param.type}({param.name});")
+            call.append(f"w_{param.name}")
+        else:
+            call.append(param.name)
+
+    return_block = "" if command.return_type == 'void' else "return"
+    
+    return TRAMPOLINE_TEMPLATE.substitute(
+        return_type=command.return_type,
+        name=command.name,
+        decl_params=command.decl_params(),
+        handle_unwrap_logic='\n'.join(handle_unwrap_logic),
+        return_block=return_block,
+        dispatch_table=dispatch_table,
+        call_params=", ".join(call),
+    )
+
+
+class EntrypointBase:
+    def __init__(self, name):
+        assert name.startswith('vk')
+        self.name = name[2:]
+        self.alias = None
+        self.guard = None
+        self.entry_table_index = None
+        # Extensions which require this entrypoint
+        self.core_version = None
+        self.extensions = []
+        self.types = None
+
+    def prefixed_name(self, prefix):
+        return prefix + '_' + self.name
+
+class Entrypoint(EntrypointBase):
+    def __init__(self, name, return_type, params):
+        super(Entrypoint, self).__init__(name)
+        self.return_type = return_type
+        self.params = params
+        self.guard = None
+        self.aliases = []
+        self.disp_table_index = None
+
+    def is_physical_device_entrypoint(self):
+        return self.params[0].type in ('VkPhysicalDevice', )
+
+    def is_device_entrypoint(self):
+        return self.params[0].type in ('VkDevice', 'VkCommandBuffer', 'VkQueue')
+
+    def decl_params(self, start=0):
+        return ', '.join(p.decl for p in self.params[start:])
+
+    def call_params(self, start=0):
+        return ', '.join(p.name for p in self.params[start:])
+
+    def is_device_dispatch(self):
+            return self.params and self.params[0].type == "VkDevice"
+
+    def is_queue_dispatch(self):
+        return self.params and self.params[0].type == "VkQueue"
+    
+    def is_command_buffer_dispatch(self):
+        return self.params and self.params[0].type == "VkCommandBuffer"
+
+    def decl_params(self):
+        return ',\n    '.join(p.decl for p in self.params)
+
+    def trampoline(self):
+        if self.alias:
+            return ""
+
+        # # Compute the types with dependencies on WrappedStructs
+        # for param in self.params:
+        #     print(generate_unwrapper(VkType(param.type)))
+
+        if self.is_physical_device_entrypoint() or self.is_device_dispatch():
+            return _generate_trampoline(self,dispatch_table="base->dispatch_table")
+        else:
+            return _generate_trampoline(self, dispatch_table="base->device->dispatch_table")
+
+class EntrypointAlias(EntrypointBase):
+    def __init__(self, name, entrypoint):
+        super(EntrypointAlias, self).__init__(name)
+        self.alias = entrypoint
+        entrypoint.aliases.append(self)
+
+    def is_physical_device_entrypoint(self):
+        return self.alias.is_physical_device_entrypoint()
+
+    def is_device_entrypoint(self):
+        return self.alias.is_device_entrypoint()
+
+    def prefixed_name(self, prefix):
+        return self.alias.prefixed_name(prefix)
+
+    @property
+    def params(self):
+        return self.alias.params
+
+    @property
+    def return_type(self):
+        return self.alias.return_type
+
+    @property
+    def disp_table_index(self):
+        return self.alias.disp_table_index
+
+    def decl_params(self):
+        return self.alias.decl_params()
+
+    def call_params(self):
+        return self.alias.call_params()
+
+def get_entrypoints(doc, api, beta):
+    """Extract the entry points from the registry."""
+    entrypoints = OrderedDict()
+
+    required = get_all_required(doc, 'command', api, beta)
+    all_types = OrderedDict()
+
+    for type in doc.findall('./types/type'):
+        if 'category' not in type.attrib or 'name' not in type.attrib or type.attrib['category'] != 'struct':
+            continue
+        name = type.attrib['name']
+        if name in SKIP:
+            continue
+        all_types[name] = type
+
+    global TYPES
+    TYPES = all_types
+
+    for t in all_types:
+        type = all_types[t]
+        if 'structextends' not in type.attrib:
+            continue
+        pnext_parents = type.attrib['structextends'].split(',')
+        for parent in pnext_parents:
+            if t in PNEXT_MEMBERS[parent]:
+                continue
+            PNEXT_MEMBERS[parent].append(t)
+
+    # Compute a fixed-point of structs with member dependencies on our WrappedStructs
+    worklist = set([t for t in WrappedStructs])
+    dependencies = defaultdict(list)
+    changed = True
+    while changed:
+        changed = False
+        for t in all_types:
+            type = all_types[t]
+            for member in type.findall('./member/type'):
+                if member.text in worklist:
+                    if t not in worklist:
+                        changed = True
+                    # Add t into worklist
+                    worklist.add(t)
+                    if member.text not in dependencies[t]:
+                        dependencies[t].append(member.text)
+
+    global TYPE_DEPENDENCIES
+    TYPE_DEPENDENCIES = dependencies
+
+    # Compute a fixed-point of structs with pnext dependencies on our WrappedStructs
+    # worklist = set([t for t in WrappedStructs])
+    dependencies = defaultdict(list)
+    changed = True
+    while changed:
+        changed = False
+        for t in all_types:
+            type = all_types[t]
+            for member in PNEXT_MEMBERS[t]:
+                if member in worklist:
+                    if t not in worklist:
+                        changed = True
+                    # Add t into worklist
+                    worklist.add(t)
+                    if member not in dependencies[t]:
+                        dependencies[t].append(member)
+
+    global PNEXT_DEPENDENCIES
+    PNEXT_DEPENDENCIES = dependencies
+
+    for command in doc.findall('./commands/command'):
+        if not filter_api(command, api):
+            continue
+
+        if 'alias' in command.attrib:
+            name = command.attrib['name']
+            target = command.attrib['alias']
+            e = EntrypointAlias(name, entrypoints[target])
+        else:
+            name = command.find('./proto/name').text
+            ret_type = command.find('./proto/type').text
+            params = [EntrypointParam(
+                type=p.find('./type').text,
+                name=p.find('./name').text,
+                decl=''.join(p.itertext()),
+                len=p.attrib.get('altlen', p.attrib.get('len', None))
+            ) for p in command.findall('./param') if filter_api(p, api)]
+            # They really need to be unique
+            e = Entrypoint(name, ret_type, params)
+
+        e.types = all_types
+
+        if name not in required:
+            continue
+
+        r = required[name]
+        e.core_version = r.core_version
+        e.extensions = r.extensions
+        e.guard = r.guard
+
+        assert name not in entrypoints, name
+        entrypoints[name] = e
+
+    return entrypoints.values()
+
+def get_entrypoints_from_xml(xml_files, beta, api='vulkan'):
+    entrypoints = []
+
+    for filename in xml_files:
+        doc = et.parse(filename)
+        entrypoints += get_entrypoints(doc, api, beta)
+
+    return entrypoints
diff --git a/src/vulkan/wrapper/vk_wrapper_trampolines_gen.py b/src/vulkan/wrapper/vk_wrapper_trampolines_gen.py
index f20fde16034..c2fa8500d98 100644
--- a/src/vulkan/wrapper/vk_wrapper_trampolines_gen.py
+++ b/src/vulkan/wrapper/vk_wrapper_trampolines_gen.py
@@ -32,7 +32,8 @@ from mako.template import Template
 
 # Mesa-local imports must be declared in meson variable
 # '{file_without_suffix}_depend_files'.
-from vk_entrypoints import get_entrypoints_from_xml
+from vk_entrypoints import get_entrypoints_from_xml, generate_unwrapper, generate_unwrapper_proto, generate_unwrapper_leaves, get_types
+import vk_entrypoints
 
 TEMPLATE_H = Template(COPYRIGHT + """\
 /* This file generated from ${filename}, don't edit directly. */
@@ -62,6 +63,16 @@ TEMPLATE_C = Template(COPYRIGHT + """\
 #include "wrapper_private.h"
 #include "wrapper_trampolines.h"
 
+${generate_unwrapper_leaves()}
+
+% for t in all_types:
+${generate_unwrapper_proto(t)}
+% endfor
+
+% for t in all_types:
+${generate_unwrapper(t)}
+% endfor
+
 % for e in entrypoints:
   % if not e.is_physical_device_entrypoint() or e.alias:
     <% continue %>
@@ -69,17 +80,7 @@ TEMPLATE_C = Template(COPYRIGHT + """\
   % if e.guard is not None:
 #ifdef ${e.guard}
   % endif
-static VKAPI_ATTR ${e.return_type} VKAPI_CALL
-${e.prefixed_name('wrapper_tramp')}(${e.decl_params()})
-{
-    <% assert e.params[0].type == 'VkPhysicalDevice' %>
-    VK_FROM_HANDLE(wrapper_physical_device, vk_physical_device, ${e.params[0].name});
-  % if e.return_type == 'void':
-    vk_physical_device->dispatch_table.${e.name}(vk_physical_device->dispatch_handle, ${e.call_params(1)});
-  % else:
-    return vk_physical_device->dispatch_table.${e.name}(vk_physical_device->dispatch_handle, ${e.call_params(1)});
-  % endif
-}
+${e.trampoline()}
   % if e.guard is not None:
 #endif
   % endif
@@ -107,58 +108,7 @@ struct vk_physical_device_entrypoint_table wrapper_physical_device_trampolines =
   % if e.guard is not None:
 #ifdef ${e.guard}
   % endif
-static VKAPI_ATTR ${e.return_type} VKAPI_CALL
-${e.prefixed_name('wrapper_tramp')}(${e.decl_params()})
-{
-  % if e.params[0].type == 'VkDevice':
-    VK_FROM_HANDLE(wrapper_device, vk_device, ${e.params[0].name});
-    % if e.return_type == 'void':
-      % if len(e.params) > 1:
-    vk_device->dispatch_table.${e.name}(vk_device->dispatch_handle, ${e.call_params(1)});
-      % else:
-    vk_device->dispatch_table.${e.name}(vk_device->dispatch_handle);
-      % endif
-    % else:
-      % if len(e.params) > 1:
-    return vk_device->dispatch_table.${e.name}(vk_device->dispatch_handle, ${e.call_params(1)});
-      % else:
-    return vk_device->dispatch_table.${e.name}(vk_device->dispatch_handle);
-      % endif
-    % endif
-  % elif e.params[0].type == 'VkCommandBuffer':
-    VK_FROM_HANDLE(wrapper_command_buffer, wcb, ${e.params[0].name});
-    % if e.return_type == 'void':
-      % if len(e.params) > 1:
-    wcb->device->dispatch_table.${e.name}(wcb->dispatch_handle, ${e.call_params(1)});
-      % else:
-    wcb->device->dispatch_table.${e.name}(wcb->dispatch_handle);
-      % endif
-    % else:
-      % if len(e.params) > 1:
-    return wcb->device->dispatch_table.${e.name}(wcb->dispatch_handle, ${e.call_params(1)});
-      % else:
-    return wcb->device->dispatch_table.${e.name}(wcb->dispatch_handle);
-      % endif
-    % endif
-  % elif e.params[0].type == 'VkQueue':
-    VK_FROM_HANDLE(wrapper_queue, wqueue, ${e.params[0].name});
-    % if e.return_type == 'void':
-      % if len(e.params) > 1:
-    wqueue->device->dispatch_table.${e.name}(wqueue->dispatch_handle, ${e.call_params(1)});
-      % else:
-    wqueue->device->dispatch_table.${e.name}(wqueue->dispatch_handle);
-      % endif
-    % else:
-      % if len(e.params) > 1:
-    return wqueue->device->dispatch_table.${e.name}(wqueue->dispatch_handle, ${e.call_params(1)});
-      % else:
-    return wqueue->device->dispatch_table.${e.name}(wqueue->dispatch_handle);
-      % endif
-    % endif
-  % else:
-    assert(!"Unhandled device child trampoline case: ${e.params[0].type}");
-  % endif
-}
+${e.trampoline()}
   % if e.guard is not None:
 #endif
   % endif
@@ -193,6 +143,7 @@ def main():
     args = parser.parse_args()
 
     entrypoints = get_entrypoints_from_xml(args.xml_files, args.beta)
+    all_types = get_types()
 
     # For outputting entrypoints.h we generate a anv_EntryPoint() prototype
     # per entry point.
@@ -200,10 +151,18 @@ def main():
         if args.out_h:
             with open(args.out_h, 'w', encoding='utf-8') as f:
                 f.write(TEMPLATE_H.render(entrypoints=entrypoints,
+                                          all_types=all_types,
+                                          generate_unwrapper_proto=generate_unwrapper_proto,
+                                          generate_unwrapper=generate_unwrapper,
+                                          generate_unwrapper_leaves=generate_unwrapper_leaves,
                                           filename=os.path.basename(__file__)))
         if args.out_c:
             with open(args.out_c, 'w', encoding='utf-8') as f:
-                f.write(TEMPLATE_C.render(entrypoints=entrypoints,
+                f.write(TEMPLATE_C.render(entrypoints=entrypoints, 
+                                          all_types=all_types,
+                                          generate_unwrapper_proto=generate_unwrapper_proto,
+                                          generate_unwrapper=generate_unwrapper,
+                                          generate_unwrapper_leaves=generate_unwrapper_leaves,
                                           filename=os.path.basename(__file__)))
     except Exception:
         # In the event there's an error, this imports some helpers from mako
diff --git a/src/vulkan/wrapper/wrapper_device.c b/src/vulkan/wrapper/wrapper_device.c
index 2ea37d1b392..6f5e559e2fb 100644
--- a/src/vulkan/wrapper/wrapper_device.c
+++ b/src/vulkan/wrapper/wrapper_device.c
@@ -138,6 +138,7 @@ wrapper_CreateDevice(VkPhysicalDevice physicalDevice,
                      const VkAllocationCallbacks* pAllocator,
                      VkDevice* pDevice)
 {
+   __log("wrapper_CreateDevice");
    VK_FROM_HANDLE(wrapper_physical_device, physical_device, physicalDevice);
    const char *wrapper_enable_extensions[VK_DEVICE_EXTENSION_COUNT];
    uint32_t wrapper_enable_extension_count = 0;
@@ -155,6 +156,11 @@ wrapper_CreateDevice(VkPhysicalDevice physicalDevice,
 
    list_inithead(&device->command_buffer_list);
    list_inithead(&device->device_memory_list);
+   list_inithead(&device->image_list);
+   list_inithead(&device->image_view_list);
+   list_inithead(&device->render_pass_list);
+   list_inithead(&device->framebuffer_list);
+
    simple_mtx_init(&device->resource_mutex, mtx_plain);
    device->physical = physical_device;
 
@@ -518,6 +524,23 @@ wrapper_DestroyDevice(VkDevice _device, const VkAllocationCallbacks* pAllocator)
                             &device->device_memory_list, link) {
       wrapper_device_memory_destroy(mem);
    }
+   list_for_each_entry_safe(struct wrapper_image, wimg,
+                            &device->image_list, link) {
+      // wrapper_image_destroy(device, wimg, pAllocator);
+   }
+   list_for_each_entry_safe(struct wrapper_image_view, wiv,
+                            &device->image_view_list, link) {
+      // wrapper_image_view_destroy(device, wiv, pAllocator);
+   }
+   list_for_each_entry_safe(struct wrapper_framebuffer, wfb,
+                          &device->framebuffer_list, link) {
+      // wrapper_framebuffer_destroy(device, wfb, pAllocator);
+   }
+   list_for_each_entry_safe(struct wrapper_render_pass, wrp,
+                          &device->render_pass_list, link) {
+      // wrapper_render_pass_destroy(device, wrp, pAllocator);
+   }
+
 
    simple_mtx_unlock(&device->resource_mutex);
 
diff --git a/src/vulkan/wrapper/wrapper_device_memory.c b/src/vulkan/wrapper/wrapper_device_memory.c
index 6d47a52fd18..1a9cb9bf068 100644
--- a/src/vulkan/wrapper/wrapper_device_memory.c
+++ b/src/vulkan/wrapper/wrapper_device_memory.c
@@ -1,10 +1,10 @@
-#define native_handle_t __native_handle_t
-#define buffer_handle_t __buffer_handle_t
+// #define native_handle_t __native_handle_t
+// #define buffer_handle_t __buffer_handle_t
 #include "wrapper_private.h"
 #include "wrapper_entrypoints.h"
 #include "vk_common_entrypoints.h"
-#undef native_handle_t
-#undef buffer_handle_t
+// #undef native_handle_t
+// #undef buffer_handle_t
 #include "util/os_file.h"
 #include "vk_util.h"
 
diff --git a/src/vulkan/wrapper/wrapper_image.c b/src/vulkan/wrapper/wrapper_image.c
new file mode 100644
index 00000000000..8b559636887
--- /dev/null
+++ b/src/vulkan/wrapper/wrapper_image.c
@@ -0,0 +1,236 @@
+#include "wrapper_private.h"
+#include "wrapper_entrypoints.h"
+#include "vk_alloc.h"
+#include "vk_common_entrypoints.h"
+#include "vk_dispatch_table.h"
+#include "vk_extensions.h"
+#include "vk_util.h"
+
+#if DETECT_OS_LINUX || DETECT_OS_BSD
+#include <drm-uapi/drm_fourcc.h>
+#endif
+
+
+static VkResult
+wrapper_image_create(struct wrapper_device *device,
+                     const VkImageCreateInfo *pCreateInfo,
+                     const VkAllocationCallbacks *pAllocator,
+                     VkImage dispatch_handle,
+                     VkImage *p_wimg)
+{
+   struct wrapper_image *wimg;
+
+   wimg = vk_object_zalloc(&device->vk, pAllocator,
+                           sizeof(struct wrapper_image),
+                           VK_OBJECT_TYPE_IMAGE);
+   if (!wimg)
+      return vk_error(&device->vk, VK_ERROR_OUT_OF_HOST_MEMORY);
+
+   vk_image_init(&device->vk, &wimg->vk, pCreateInfo);
+
+   wimg->device = device;
+   wimg->dispatch_handle = dispatch_handle;
+   list_add(&wimg->link, &device->image_list);
+
+   *p_wimg = wrapper_image_to_handle(wimg);
+
+   return VK_SUCCESS;
+}
+
+void
+wrapper_image_destroy(struct wrapper_device *device,
+                      struct wrapper_image *wimg,
+                      const VkAllocationCallbacks* pAllocator)
+{
+   list_del(&wimg->link);
+   vk_image_finish(&wimg->vk);
+   vk_image_destroy(&device->vk, pAllocator, &wimg->vk);
+}
+
+VKAPI_ATTR VkResult VKAPI_CALL
+wrapper_CreateImage(VkDevice _device,
+                    const VkImageCreateInfo* pCreateInfo,
+                    const VkAllocationCallbacks* pAllocator,
+                    VkImage* pImage)
+{
+   VK_FROM_HANDLE(wrapper_device, device, _device);
+   VkImageCreateInfo create_info = *pCreateInfo;
+   VkImage dispatch_handle;
+   VkResult result;
+
+   /* The wrapper may need to perform blits or copies. */
+   create_info.usage |= VK_IMAGE_USAGE_TRANSFER_SRC_BIT |
+                        VK_IMAGE_USAGE_TRANSFER_DST_BIT;
+
+   result = device->dispatch_table.CreateImage(device->dispatch_handle,
+                                               &create_info,
+                                               pAllocator,
+                                               &dispatch_handle);
+   if (result != VK_SUCCESS)
+      return result;
+
+   simple_mtx_lock(&device->resource_mutex);
+
+   result = wrapper_image_create(device, &create_info, pAllocator,
+                                 dispatch_handle, pImage);
+
+   if (result != VK_SUCCESS) {
+      device->dispatch_table.DestroyImage(device->dispatch_handle,
+                                          dispatch_handle, pAllocator);
+   }
+
+   simple_mtx_unlock(&device->resource_mutex);
+
+   return result;
+}
+
+VKAPI_ATTR void VKAPI_CALL
+wrapper_DestroyImage(VkDevice _device,
+                     VkImage _image,
+                     const VkAllocationCallbacks* pAllocator)
+{
+   VK_FROM_HANDLE(wrapper_device, device, _device);
+   VK_FROM_HANDLE(wrapper_image, wimg, _image);
+
+   if (!_image)
+       return;
+
+   device->dispatch_table.DestroyImage(device->dispatch_handle,
+                                       wimg->dispatch_handle, pAllocator);
+
+   simple_mtx_lock(&device->resource_mutex);
+   wrapper_image_destroy(device, wimg, pAllocator);
+   simple_mtx_unlock(&device->resource_mutex);
+}
+
+// #if DETECT_OS_LINUX || DETECT_OS_BSD
+// VKAPI_ATTR VkResult VKAPI_CALL
+// wrapper_GetImageDrmFormatModifierPropertiesEXT(VkDevice _device,
+//                                                VkImage _image,
+//                                                VkImageDrmFormatModifierPropertiesEXT* pProperties)
+// {
+//    VK_FROM_HANDLE(wrapper_device, device, _device);
+//    VK_FROM_HANDLE(wrapper_image, wimg, _image);
+
+//    return device->dispatch_table.GetImageDrmFormatModifierPropertiesEXT(
+//       device->dispatch_handle, wimg->dispatch_handle, pProperties);
+// }
+// #endif
+
+// VKAPI_ATTR void VKAPI_CALL
+// wrapper_GetImageSubresourceLayout(VkDevice _device,
+//                                   VkImage _image,
+//                                   const VkImageSubresource* pSubresource,
+//                                   VkSubresourceLayout* pLayout)
+// {
+//    VK_FROM_HANDLE(wrapper_device, device, _device);
+//    VK_FROM_HANDLE(wrapper_image, wimg, _image);
+
+//    device->dispatch_table.GetImageSubresourceLayout(device->dispatch_handle,
+//                                                     wimg->dispatch_handle,
+//                                                     pSubresource,
+//                                                     pLayout);
+// }
+
+// VKAPI_ATTR void VKAPI_CALL
+// wrapper_GetImageMemoryRequirements(VkDevice _device,
+//                                    VkImage _image,
+//                                    VkMemoryRequirements* pMemoryRequirements)
+// {
+//    VK_FROM_HANDLE(wrapper_device, device, _device);
+//    VK_FROM_HANDLE(wrapper_image, wimg, _image);
+
+//    device->dispatch_table.GetImageMemoryRequirements(device->dispatch_handle,
+//                                                      wimg->dispatch_handle,
+//                                                      pMemoryRequirements);
+// }
+
+// VKAPI_ATTR void VKAPI_CALL
+// wrapper_GetImageMemoryRequirements2(VkDevice _device,
+//                                     const VkImageMemoryRequirementsInfo2* pInfo,
+//                                     VkMemoryRequirements2* pMemoryRequirements)
+// {
+//    VK_FROM_HANDLE(wrapper_device, device, _device);
+//    VK_FROM_HANDLE(wrapper_image, wimg, pInfo->image);
+//    VkImageMemoryRequirementsInfo2 info = *pInfo;
+//    info.image = wimg->dispatch_handle;
+
+//    device->dispatch_table.GetImageMemoryRequirements2(device->dispatch_handle,
+//                                                       &info,
+//                                                       pMemoryRequirements);
+// }
+
+// VKAPI_ATTR VkResult VKAPI_CALL
+// wrapper_BindImageMemory(VkDevice _device,
+//                         VkImage _image,
+//                         VkDeviceMemory memory,
+//                         VkDeviceSize memoryOffset)
+// {
+//    VK_FROM_HANDLE(wrapper_device, device, _device);
+//    VK_FROM_HANDLE(wrapper_image, wimg, _image);
+
+//    return device->dispatch_table.BindImageMemory(device->dispatch_handle,
+//                                                   wimg->dispatch_handle,
+//                                                   memory,
+//                                                   memoryOffset);
+// }
+
+// VKAPI_ATTR VkResult VKAPI_CALL
+// wrapper_BindImageMemory2(VkDevice _device,
+//                          uint32_t bindInfoCount,
+//                          const VkBindImageMemoryInfo* pBindInfos)
+// {
+//    VK_FROM_HANDLE(wrapper_device, device, _device);
+//    STACK_ARRAY(VkBindImageMemoryInfo, unwrapped_infos, bindInfoCount);
+//    if (!unwrapped_infos)
+//       return vk_error(&device->vk, VK_ERROR_OUT_OF_HOST_MEMORY);
+
+//    for (uint32_t i = 0; i < bindInfoCount; i++) {
+//       unwrapped_infos[i] = pBindInfos[i];
+//       VK_FROM_HANDLE(wrapper_image, wimg, pBindInfos[i].image);
+//       unwrapped_infos[i].image = wimg->dispatch_handle;
+//    }
+
+//    VkResult result = device->dispatch_table.BindImageMemory2(device->dispatch_handle,
+//                                                              bindInfoCount,
+//                                                              unwrapped_infos);
+
+//    STACK_ARRAY_FINISH(unwrapped_infos);
+//    return result;
+// }
+
+// VKAPI_ATTR void VKAPI_CALL
+// wrapper_GetImageSparseMemoryRequirements(
+//    VkDevice _device,
+//    VkImage _image,
+//    uint32_t* pSparseMemoryRequirementCount,
+//    VkSparseImageMemoryRequirements* pSparseMemoryRequirements)
+// {
+//    VK_FROM_HANDLE(wrapper_device, device, _device);
+//    VK_FROM_HANDLE(wrapper_image, wimg, _image);
+
+//    device->dispatch_table.GetImageSparseMemoryRequirements(
+//       device->dispatch_handle,
+//       wimg->dispatch_handle,
+//       pSparseMemoryRequirementCount,
+//       pSparseMemoryRequirements);
+// }
+
+// VKAPI_ATTR void VKAPI_CALL
+// wrapper_GetImageSparseMemoryRequirements2(
+//    VkDevice _device,
+//    const VkImageSparseMemoryRequirementsInfo2* pInfo,
+//    uint32_t* pSparseMemoryRequirementCount,
+//    VkSparseImageMemoryRequirements2* pSparseMemoryRequirements)
+// {
+//    VK_FROM_HANDLE(wrapper_device, device, _device);
+//    VK_FROM_HANDLE(wrapper_image, wimg, pInfo->image);
+//    VkImageSparseMemoryRequirementsInfo2 info = *pInfo;
+//    info.image = wimg->dispatch_handle;
+
+//    device->dispatch_table.GetImageSparseMemoryRequirements2(
+//       device->dispatch_handle,
+//       &info,
+//       pSparseMemoryRequirementCount,
+//       pSparseMemoryRequirements);
+// }
diff --git a/src/vulkan/wrapper/wrapper_image_view.c b/src/vulkan/wrapper/wrapper_image_view.c
new file mode 100644
index 00000000000..52497ab384a
--- /dev/null
+++ b/src/vulkan/wrapper/wrapper_image_view.c
@@ -0,0 +1,124 @@
+#include "wrapper_private.h"
+#include "wrapper_entrypoints.h"
+#include "vk_alloc.h"
+#include "vk_common_entrypoints.h"
+#include "vk_dispatch_table.h"
+#include "vk_extensions.h"
+#include "vk_image.h"
+// #include "vk_image_view.h"
+#include "vk_util.h"
+
+// static VkResult
+// wrapper_image_view_create(struct wrapper_device *device,
+//                           const VkImageViewCreateInfo* pCreateInfo,
+//                           const VkAllocationCallbacks* pAllocator,
+//                           VkImageView dispatch_handle,
+//                           VkImageView* p_wiv)
+// {
+//    struct wrapper_image_view *wiv;
+
+//    wiv = vk_object_zalloc(&device->vk, pAllocator,
+//                           sizeof(struct wrapper_image_view),
+//                           VK_OBJECT_TYPE_IMAGE_VIEW);
+//    if (!wiv)
+//       return vk_error(&device->vk, VK_ERROR_OUT_OF_HOST_MEMORY);
+
+//    vk_image_view_init(&device->vk, &wiv->vk, false, pCreateInfo);
+
+//    wiv->device = device;
+//    wiv->dispatch_handle = dispatch_handle;
+//    list_add(&wiv->link, &device->image_view_list);
+
+//    *p_wiv = wrapper_image_view_to_handle(wiv);
+
+//    return VK_SUCCESS;
+// }
+
+// void
+// wrapper_image_view_destroy(struct wrapper_device *device,
+//                            struct wrapper_image_view *wiv,
+//                            const VkAllocationCallbacks* pAllocator)
+// {
+//    list_del(&wiv->link);
+//    vk_image_view_finish(&wiv->vk);
+//    vk_image_view_destroy(&device->vk, pAllocator, &wiv->vk);
+// }
+
+
+// VKAPI_ATTR VkResult VKAPI_CALL
+// wrapper_CreateImageView(VkDevice _device,
+//                         const VkImageViewCreateInfo* pCreateInfo,
+//                         const VkAllocationCallbacks* pAllocator,
+//                         VkImageView* pView)
+// {
+//    VK_FROM_HANDLE(wrapper_device, device, _device);
+//    VK_FROM_HANDLE(wrapper_image, wimg, pCreateInfo->image);
+
+//    VkImageViewCreateInfo create_info = *pCreateInfo;
+//    create_info.image = wimg->dispatch_handle;
+
+//    /* The wrapper may need to blit to/from this image view, so we need
+//     * to ensure the view's usage includes the transfer bits.
+//     */
+//    VkImageViewUsageCreateInfo usage_info;
+//    VkImageViewUsageCreateInfo *usage_info_ptr =
+//       (VkImageViewUsageCreateInfo *) vk_find_struct(
+//          (void *)pCreateInfo, IMAGE_VIEW_USAGE_CREATE_INFO);
+
+//    if (usage_info_ptr) {
+//       /* The app provided one, so we need to add our bits. */
+//       usage_info = *usage_info_ptr;
+//    } else {
+//       /* No usage info provided, so the view inherits the image's usage.
+//        * Since we added the transfer bits to the image, that's sufficient.
+//        * However, if we're creating a view of a specific format that's
+//        * different from the image's format, we might need to add our own.
+//        * Let's just add one to be safe. */
+//       usage_info = (VkImageViewUsageCreateInfo) {
+//          .sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_USAGE_CREATE_INFO,
+//          .usage = wimg->vk.usage,
+//       };
+//    }
+
+//    usage_info.usage |= VK_IMAGE_USAGE_TRANSFER_SRC_BIT |
+//                        VK_IMAGE_USAGE_TRANSFER_DST_BIT;
+//    __vk_append_struct(&create_info, &usage_info);
+
+//    VkImageView dispatch_handle;
+//    VkResult result = device->dispatch_table.CreateImageView(device->dispatch_handle,
+//                                                             &create_info,
+//                                                             pAllocator,
+//                                                             &dispatch_handle);
+//    if (result != VK_SUCCESS)
+//       return result;
+
+//    simple_mtx_lock(&device->resource_mutex);
+//    result = wrapper_image_view_create(device, pCreateInfo, pAllocator,
+//                                       dispatch_handle, pView);
+//    if (result != VK_SUCCESS) {
+//       device->dispatch_table.DestroyImageView(device->dispatch_handle,
+//                                               dispatch_handle, pAllocator);
+//    }
+//    simple_mtx_unlock(&device->resource_mutex);
+
+//    return result;
+// }
+
+// VKAPI_ATTR void VKAPI_CALL
+// wrapper_DestroyImageView(VkDevice _device,
+//                          VkImageView imageView,
+//                          const VkAllocationCallbacks* pAllocator)
+// {
+//    VK_FROM_HANDLE(wrapper_device, device, _device);
+//    VK_FROM_HANDLE(wrapper_image_view, wiv, imageView);
+
+//    if (!imageView)
+//       return;
+
+//    device->dispatch_table.DestroyImageView(device->dispatch_handle,
+//                                            wiv->dispatch_handle, pAllocator);
+
+//    simple_mtx_lock(&device->resource_mutex);
+//    wrapper_image_view_destroy(device, wiv, pAllocator);
+//    simple_mtx_unlock(&device->resource_mutex);
+// }
diff --git a/src/vulkan/wrapper/wrapper_instance.c b/src/vulkan/wrapper/wrapper_instance.c
index 237108ca6eb..0162bc7bc6d 100644
--- a/src/vulkan/wrapper/wrapper_instance.c
+++ b/src/vulkan/wrapper/wrapper_instance.c
@@ -279,6 +279,7 @@ VKAPI_ATTR PFN_vkVoidFunction VKAPI_CALL
 wrapper_GetInstanceProcAddr(VkInstance _instance,
                             const char *pName)
 {
+
    VK_FROM_HANDLE(wrapper_instance, instance, _instance);
    return vk_instance_get_proc_addr(&instance->vk,
                                     &wrapper_instance_entrypoints,
diff --git a/src/vulkan/wrapper/wrapper_physical_device.c b/src/vulkan/wrapper/wrapper_physical_device.c
index 384c45483c6..0f9cd419c71 100644
--- a/src/vulkan/wrapper/wrapper_physical_device.c
+++ b/src/vulkan/wrapper/wrapper_physical_device.c
@@ -12,8 +12,16 @@
 #include "wsi_common.h"
 #include "util/os_misc.h"
 
+int __android_log_print(
+  int prio,
+  const char *tag,
+  const char *fmt,
+  ...
+);
+
 static VkResult
 wrapper_setup_device_extensions(struct wrapper_physical_device *pdevice) {
+   __android_log_print(6, "Wrapper", "inside wrapper_setup_device_extensions");
    struct vk_device_extension_table *exts = &pdevice->vk.supported_extensions;
    VkExtensionProperties pdevice_extensions[VK_DEVICE_EXTENSION_COUNT];
    uint32_t pdevice_extension_count = VK_DEVICE_EXTENSION_COUNT;
@@ -466,9 +474,11 @@ wrapper_GetPhysicalDeviceFormatProperties(VkPhysicalDevice physicalDevice,
                                             VkFormatProperties* pFormatProperties)
 {
    VK_FROM_HANDLE(wrapper_physical_device, pdevice, physicalDevice);
-   pdevice->dispatch_table.GetPhysicalDeviceFormatProperties(
-      pdevice->dispatch_handle, format, pFormatProperties);
-      
+   struct vk_features *supported_features = &pdevice->base_supported_features;
+
+   VkFormat targetFormat = format;
+   bool modified = false;
+
    switch (format) {
    case VK_FORMAT_BC1_RGB_SRGB_BLOCK:
    case VK_FORMAT_BC1_RGB_UNORM_BLOCK:
@@ -486,9 +496,17 @@ wrapper_GetPhysicalDeviceFormatProperties(VkPhysicalDevice physicalDevice,
    case VK_FORMAT_BC6H_SFLOAT_BLOCK:
    case VK_FORMAT_BC7_UNORM_BLOCK:
    case VK_FORMAT_BC7_SRGB_BLOCK:
-      pFormatProperties->optimalTilingFeatures |= VK_FORMAT_FEATURE_SAMPLED_IMAGE_BIT | VK_FORMAT_FEATURE_BLIT_SRC_BIT | VK_FORMAT_FEATURE_SAMPLED_IMAGE_FILTER_LINEAR_BIT | VK_FORMAT_FEATURE_TRANSFER_DST_BIT;
+      if (!supported_features->textureCompressionBC) {
+         // Decompress everything into VK_FORMAT_R8G8B8A8_UNORM
+         targetFormat = VK_FORMAT_R8G8B8A8_UNORM;
+         modified = true;
+      }
       break;
    default:
       break;   
    }
+   pdevice->dispatch_table.GetPhysicalDeviceFormatProperties(pdevice->dispatch_handle, targetFormat, pFormatProperties);
+   if (modified) {
+      pFormatProperties->optimalTilingFeatures |= VK_FORMAT_FEATURE_SAMPLED_IMAGE_BIT | VK_FORMAT_FEATURE_BLIT_SRC_BIT | VK_FORMAT_FEATURE_SAMPLED_IMAGE_FILTER_LINEAR_BIT | VK_FORMAT_FEATURE_TRANSFER_DST_BIT;
+   }
 }                                      
diff --git a/src/vulkan/wrapper/wrapper_private.h b/src/vulkan/wrapper/wrapper_private.h
index 7f8cc6e8c84..843e0f7a1e5 100644
--- a/src/vulkan/wrapper/wrapper_private.h
+++ b/src/vulkan/wrapper/wrapper_private.h
@@ -5,7 +5,11 @@
 #include "vulkan/runtime/vk_device.h"
 #include "vulkan/runtime/vk_queue.h"
 #include "vulkan/runtime/vk_command_buffer.h"
+#include "vulkan/runtime/vk_image.h"
 #include "vulkan/runtime/vk_log.h"
+#include "vulkan/runtime/vk_render_pass.h"
+#include "vulkan/runtime/vk_framebuffer.h"
+
 #include "vulkan/util/vk_dispatch_table.h"
 #include "vulkan/wsi/wsi_common.h"
 #include "util/simple_mtx.h"
@@ -15,6 +19,24 @@ extern const struct vk_instance_extension_table wrapper_instance_extensions;
 extern const struct vk_device_extension_table wrapper_device_extensions;
 extern const struct vk_device_extension_table wrapper_filter_extensions;
 
+static void __log(const char* fmt, ...) {
+   static FILE* fd;
+   static pid_t pid;
+   if (fd == NULL) {
+      pid = getpid();
+      char buf[256];
+      sprintf(buf, "/sdcard/Documents/Wrapper/wrapper.txt");
+      fd = fopen(buf, "w");
+   }
+   // fprintf(fd, "[%d] ", pid);
+   va_list args;
+   va_start(args, fmt);
+   vfprintf(fd, fmt, args);
+   va_end(args);
+   fprintf(fd, "\n");
+   fflush(fd);
+}
+
 struct wrapper_instance {
    struct vk_instance vk;
 
@@ -59,7 +81,12 @@ struct wrapper_device {
    VkDevice dispatch_handle;
    simple_mtx_t resource_mutex;
    struct list_head command_buffer_list;
+   struct list_head image_list;
+   struct list_head image_view_list;
    struct list_head device_memory_list;
+   struct list_head render_pass_list;
+   struct list_head framebuffer_list;
+
    struct wrapper_physical_device *physical;
    struct vk_device_dispatch_table dispatch_table;
 };
@@ -79,6 +106,52 @@ struct wrapper_command_buffer {
 VK_DEFINE_HANDLE_CASTS(wrapper_command_buffer, vk.base, VkCommandBuffer,
                        VK_OBJECT_TYPE_COMMAND_BUFFER)
 
+struct wrapper_image {
+   struct vk_image vk;
+
+   struct wrapper_device *device;
+   struct list_head link;
+   VkImage dispatch_handle;
+
+   VkFormat original_format;
+};
+
+VK_DEFINE_NONDISP_HANDLE_CASTS(wrapper_image, vk.base, VkImage, VK_OBJECT_TYPE_IMAGE)
+
+struct wrapper_image_view {
+       struct vk_image_view vk;
+
+       struct wrapper_device *device;
+       struct list_head link;
+       VkImageView dispatch_handle;
+};
+
+VK_DEFINE_NONDISP_HANDLE_CASTS(wrapper_image_view, vk.base, VkImageView,
+                               VK_OBJECT_TYPE_IMAGE_VIEW)
+
+struct wrapper_render_pass {
+       struct vk_render_pass vk;
+
+       struct wrapper_device *device;
+       struct list_head link;
+       VkRenderPass dispatch_handle;
+};
+
+VK_DEFINE_NONDISP_HANDLE_CASTS(wrapper_render_pass, vk.base, VkRenderPass,
+                               VK_OBJECT_TYPE_RENDER_PASS)
+
+struct wrapper_framebuffer {
+       struct vk_object_base vk;
+
+       struct wrapper_device *device;
+       struct list_head link;
+       VkFramebuffer dispatch_handle;
+};
+
+VK_DEFINE_NONDISP_HANDLE_CASTS(wrapper_framebuffer, vk, VkFramebuffer,
+                               VK_OBJECT_TYPE_FRAMEBUFFER)
+
+
 struct wrapper_device_memory {
    struct AHardwareBuffer *ahardware_buffer;
    struct wrapper_device *device;
@@ -108,3 +181,23 @@ wrapper_device_memory_create(struct wrapper_device *device,
 
 void
 wrapper_device_memory_destroy(struct wrapper_device_memory *mem);
+
+void
+wrapper_image_destroy(struct wrapper_device *device,
+                      struct wrapper_image *wimg,
+                      const VkAllocationCallbacks *pAllocator);
+
+void
+wrapper_image_view_destroy(struct wrapper_device *device,
+                           struct wrapper_image_view *wiv,
+                           const VkAllocationCallbacks *pAllocator);
+
+void
+wrapper_render_pass_destroy(struct wrapper_device *device,
+                            struct wrapper_render_pass *wrp,
+                            const VkAllocationCallbacks *pAllocator);
+
+void
+wrapper_framebuffer_destroy(struct wrapper_device *device,
+                            struct wrapper_framebuffer *wfb,
+                            const VkAllocationCallbacks *pAllocator);
diff --git a/src/vulkan/wsi/wsi_common_x11.c b/src/vulkan/wsi/wsi_common_x11.c
index 51a11b875ad..39b1faa18b7 100644
--- a/src/vulkan/wsi/wsi_common_x11.c
+++ b/src/vulkan/wsi/wsi_common_x11.c
@@ -64,13 +64,13 @@
 
 #include <android/hardware_buffer.h>
 
-typedef struct
-{
-    int version;        /* sizeof(native_handle) */
-    int numFds;         /* number of file-descriptors at &data[0] */
-    int numInts;        /* number of ints at &data[numFds] */
-    int data[0];        /* numFds + numInts ints */
-} native_handle_t;
+// typedef struct native_handle
+// {
+//     int version;        /* sizeof(native_handle) */
+//     int numFds;         /* number of file-descriptors at &data[0] */
+//     int numInts;        /* number of ints at &data[numFds] */
+//     int data[0];        /* numFds + numInts ints */
+// } native_handle_t;
 
 const native_handle_t* _Nullable AHardwareBuffer_getNativeHandle(
         const AHardwareBuffer* _Nonnull buffer);
